### PR TITLE
GH-1491: Support Optional/null Payloads

### PR DESCRIPTION
**Pull request by garyrussell**
Resolves https://github.com/spring-projects/spring-amqp/issues/1491

**cherry-pick to 2.4.x**

### CODE REVIEW

File: spring-rabbit/src/test/java/org/springframework/amqp/rabbit/annotation/OptionalPayloadTests.java [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+
+		@Bean
+		ConnectionFactory rabbitConnectionFactory() {
+			BrokerRunningSupport brokerRunning = RabbitAvailableCondition.getBrokerRunning();
+			CachingConnectionFactory connectionFactory = new CachingConnectionFactory();
+			connectionFactory.setHost(brokerRunning.getHostName());
+			connectionFactory.setPort(brokerRunning.getPort());
+			connectionFactory.setUsername(brokerRunning.getUser());
+			connectionFactory.setPassword(brokerRunning.getPassword());
+			return connectionFactory;
```
Comment (artembilan) [inline spring-rabbit/src/test/java/org/springframework/amqp/rabbit/annotation/OptionalPayloadTests.java:112]: How about this instead: `return new CachingConnectionFactory(RabbitAvailableCondition.getBrokerRunning().getConnectionFactory());` ?

File: src/reference/asciidoc/amqp.adoc [hunk 2]
```diff
@@ -4081,6 +4081,38 @@ converter to determine the type.
...
 By default and for backward compatibility, the `Jackson2JsonMessageConverter` trusts all packages -- that is, it uses `*` for the option.
 
+Starting with version 2.4.7, the converter can be configured to return `Optional.empty()` if Jackson returns `null` after deserializing the message body.
+This facilitates `@RabbitListener` s to receive null payloads, in two ways:
+
+====
+[source, java]
+----
+@RabbitListener(queues = "op.1")
+void listen(@Payload(required = false) Thing payload) {
```
Comment (artembilan) [inline src/reference/asciidoc/amqp.adoc:4091]: I wonder if `PayloadMethodArgumentResolver` should treat a `@Nullable` as equal to this one.
Much shorter and intuitive for end-user codding 😄

**Review by artembilan — CHANGES_REQUESTED** [review]

File: spring-rabbit/src/test/java/org/springframework/amqp/rabbit/annotation/OptionalPayloadTests.java (hunk 1 above)
Comment (garyrussell) [reply spring-rabbit/src/test/java/org/springframework/amqp/rabbit/annotation/OptionalPayloadTests.java:112]: Yeah copy/paste from another test that was exercising these methods.

File: src/reference/asciidoc/amqp.adoc (hunk 2 above)
Comment (garyrussell) [reply src/reference/asciidoc/amqp.adoc:4091]: True; feel free to open an issue against Framework :D

Comment (artembilan) [reply src/reference/asciidoc/amqp.adoc:4091]: Here you are: https://github.com/spring-projects/spring-framework/issues/28945 !

**Comment by artembilan** [comment]
Back-ported as https://github.com/spring-projects/spring-amqp/commit/3d3dfa5d70adc05a6f774fcb32f45e9f24a28e8b after fixing conflict in the `whats-new.adoc`

**Pull request merged**